src/utils.ts

```diff
@@ -182,7 +183,123 @@ export function interpolate(
   });
 }
 
+export const bigIntify = (val: any) => BigInt(val);
+
+export const bigNumberify = (val: any) => new BigNumber(val);
+
+export const stringify = (val: any) => val.toString();
+
+export const catchParseLogError = (e: any, logger: Logger) => {
+  if (e instanceof Error) {
+    if (!e.message.includes('no matching event')) {
+      logger.error('Failed parse event', e);
+    }
+  }
+};
+
+const PREFIX_BIG_INT = 'bi@';
+const PREFIX_BIG_NUMBER = 'bn@';
+
+const stringCheckerBuilder = (prefix: string) => {
+  return (obj: any) => {
+    if (!_.isString(obj)) {
+      return false;
+    }
+    for (let i = 0; i < prefix.length; ++i) {
+      if (prefix[i] !== obj[i]) {
+        return false;
+      }
+    }
+    return true;
+  };
+};
+
+const checkerStringWithBigIntPrefix = stringCheckerBuilder(PREFIX_BIG_INT);
+const checkerStringWithBigNumberPrefix =
+  stringCheckerBuilder(PREFIX_BIG_NUMBER);
+
+const casterToStringbuilder = (prefix: string, obj: any) =>
+  prefix.concat(obj.toString());
+
+const casterBigIntToString = (obj: BigInt) =>
+  casterToStringbuilder(PREFIX_BIG_INT, obj);
+const casterBigNumberToString = (obj: BigNumber) =>
+  casterToStringbuilder(PREFIX_BIG_NUMBER, obj);
+
+const checkerBigInt = (obj: any) => typeof obj === 'bigint';
+const checkerBigNumber = (obj: any) => obj instanceof BigNumber;
+
+const stringCasterBuilder = (
+  prefix: string,
+  constructor: (str: string) => any,
+) => {
+  return (obj: string) => {
+    return constructor(obj.slice(prefix.length));
+  };
+};
+
+const casterStringToBigInt = stringCasterBuilder(
+  PREFIX_BIG_INT,
+  (str: string) => BigInt(str),
+);
+
+const casterStringToBigNumber = stringCasterBuilder(
+  PREFIX_BIG_NUMBER,
+  (str: string) => new BigNumber(str),
+);
+
+type TypeSerializer = {
+  checker: (obj: any) => boolean;
+  caster: (obj: any) => any;
+};
+
+export function deepTypecast(obj: any, types: TypeSerializer[]): any {
```

Comment:
Well, maybe typing so many things will lead to problems and maybe not worth it. Or we can try :)